### PR TITLE
RAC-5431 vmslave booking && recovery

### PR DESCRIPTION
**Pull request by changev**
add vmslave booking recovery system to keep sandbox vmslave safety.

docs: https://hardware.corp.emc.com/display/gheonrack/SH+Jenkins+Sandbox

@panpan0000 @PengTian0

### CODE REVIEW

**Comment by panpan0000** [comment]
@changev , I believe you are doing RAC-5431. please add this in PR title.

**Comment by panpan0000** [comment]
@changev , Wrong rebase ....
your PR contains 2 of my commit..

File: deployment/vm_operation.sh [hunk 1]
```diff
@@ -67,6 +67,58 @@ power_action() #id_string action(on/off/reset)
...
+        vm_name=`vm_getname $id`
+
+        if [ $snapshot_count -gt 1 ]; then
+            echo "Snapshots of $vm_name is in bad status, please contact administrator."
+            exit 1
+        fi
+
+        if [ "$action" = "create" ]; then
+            if [ $snapshot_count = 1 ]; then
+                echo "Snapshot creation failed, snapshot of this vm already exists."
```
Comment (panpan0000) [inline deployment/vm_operation.sh]: please add ```$vm_name```  in this error message

File: deployment/vm_operation.sh [hunk 2]
```diff
@@ -67,6 +67,58 @@ power_action() #id_string action(on/off/reset)
...
+{
+    action=$1
+    duration=$2
+    shift 2
+    for id in $@
+    do
+        snapshot_count=`vim-cmd vmsvc/snapshot.get $id | grep "Snapshot Id"| wc -l`
+        vm_name=`vm_getname $id`
+
+        if [ $snapshot_count -gt 1 ]; then
```
Comment (panpan0000) [inline deployment/vm_operation.sh:80]: When the vmslave ansible updated, the vmslave will be re-configure, and the snapshot will be re-created.
so is it reasonable to only allow only one snapshot in total ?
when we revert to snapshot, Can we always revert to latest/last snapshot ?
```
   last_snapshot_id=$(  vim-cmd  vmsvc/snapshot.get ${vmid}|   grep "Snapshot Id"|  tail -1|awk '{print $NF}')
   last_snapshot_name=$(vim-cmd  vmsvc/snapshot.get ${vmid}| grep "Snapshot Name"|  tail -1|awk '{print $NF}')
```

File: deployment/vm_operation.sh [hunk 3]
```diff
@@ -67,6 +67,58 @@ power_action() #id_string action(on/off/reset)
...
+            if [ $? -ne 0 ];then
+                echo "ERROR: snapshot_create VM $vm_name fails"
+                continue
+            fi
+        fi
+
+        snapshot_id=`vim-cmd vmsvc/snapshot.get $id | grep "Snapshot Id" | awk '{print $4}'`
+        if [ "$action" = "revert" ]; then
+            if [ $snapshot_count = 0 ]; then
+                echo "Revert snapshot failed, there's no snapshot of this vm."
```
Comment (panpan0000) [inline deployment/vm_operation.sh]: ```echo "Revert snapshot failed, there's no snapshot of vm: $vm_name"```
adding $vm_name

File: deployment/vm_operation.sh [hunk 4]
```diff
@@ -67,6 +67,58 @@ power_action() #id_string action(on/off/reset)
...
+                echo "If this is not expected, please contact administrator."
+                exit 1
+            fi
+            vim-cmd vmsvc/snapshot.revert $id $snapshot_id false
+            if [ $? -ne 0 ];then
+                echo "ERROR: snapshot_revert VM $vm_name fails"
+                continue
+            fi
+            vim-cmd vmsvc/power.on $id 2>&1 > /dev/null
+            vim-cmd vmsvc/snapshot.remove $id $snapshot_id
```
Comment (panpan0000) [inline deployment/vm_operation.sh]: why remove the snapshot ?

**Comment by panpan0000** [comment]
General comments:

your solution sounds workable if people follow the process to book/unbook the vmslave if he/she is aware of the impact of change he/she will make.

but how about this situation ?  someone create a script/pipeline which has destructive impact to vmslaves  but he/she is not well aware of that, so he/she didn't follow the book/unbook procedure.
example , "change network setting in a new script" .  then the impact to the vmslave will persist .

following your idea, I was thinking of taking snapshot after vmslave being configured/re-config, and this/those snapshot(s) can be used to restored if a destructive impact happens ? ( the negative consequence will be the vmslave will lost the workspace data it had)

File: deployment/vm_operation.sh (hunk 2 above)
Comment (changev) [reply deployment/vm_operation.sh:80]: 1. If we use this book system, then global updates won't be done if any vmslave is booked. If updates is necessary, we should disscuss with the developer or we regard this booked slave as an exception, we can update it after booking cancel.

2. In my proposalthere won't be more than one snapshot of a vm.
If so something must go wrong and throw error at the first time is necessary.
Use latest snapshot  may cover such issues.

Comment (changev) [reply deployment/vm_operation.sh:80]: in summary, IMO keep only one snapshot is more robust and easier to manage.

**Comment by changev** [comment]
@panpan0000 Thanks for your advice.

Keep a resident snapshot of every vm is good for security. But it increase the spending.
Every maintainer(now me) must know this external infomation, they must delete the old resident  snapshot and create a new one(maybe use a jenkins job) after every time vmslave configurating.
This is a long time work.

In other words, "don't follow the book/unbook procedure and destroy a vmslave" may be rare case.

However, IMO this is a cost/benefit tradeoff. Both is OK to me.

**Review by panpan0000 — APPROVED** [review]

**Comment by panpan0000** [comment]
Ok.
there's 2 cases:
(1) in someone's PR, the modification of CI scripts will do destructive impact to vmslave. and when this PR goes into PR Gate, the harm will be done.
(2) people books a vmslave, and do any change, then some automation helps him to restore the vmslave .

@changev's PR address the point 2.
point (1) is hard to do. will not be in short term scope.

**Review by iceiilin — APPROVED** [review]

**Pull request merged**